### ISSUE TITLE
Test SQL migration down scripts

### ISSUE THREAD

**Issue by tgeoghegan** [divviup/janus#1279]
Test setups will apply SQL migration up scripts against ephemeral databases, and in #1277, we run tests against multiple schema versions. However we don't currently have coverage of down scripts. We should add tests that run downgrade scripts and as much as possible, run database tests against downgraded databases.

**Comment by tgeoghegan** [comment]
There's a few different things we can check here:

- that down scripts can run at all
- that down scripts return the schema to the state it was in before running the corresponding up script (perhaps by dumping the schema before and after the up/down migration pair and diffing?)
- that Janus can run against a database that has been downgraded (this is tricky because the Janus version under test doesn't necessary work against all the historical database versions, but we can at least do this for schema versions in `datastore::SUPPORTED_SCHEMA_VERSIONS`

**Comment by inahga** [comment]
I was able to write a test that verifies whether the scripts can run :+1:.

As for schema validation, there's a few options:
1. Dump tables/indices/types/etc. by hand and compare.
2. Find a dedicated tool for diff'ing pg schemas.

The latter sounds like the better idea--but I'm still researching what the best options are. Open to input if anyone knows a good tool off-hand.

> that Janus can run against a database that has been downgraded (this is tricky because the Janus version under test doesn't necessary work against all the historical database versions, but we can at least do this for schema versions in datastore::SUPPORTED_SCHEMA_VERSIONS

I lean towards making this an integration test for every `datastore::SUPPORTED_SCHEMA_VERSIONS`, and extend tests to allow running them on a particular schema version. I don't think there's a clean way to re-invoke the whole test suite from within the test-suite. It might also call for extending tests to indicate their own minimum schema version, to accommodate for tests that exercise newer versions of the schema. Might get very tricky...

**Comment by divergentdave** [comment]
Currently, we can re-run datastore tests against multiple schema versions (only using `.up.sql` scripts) using an rstest_reuse template. We have a macro that takes in a variadic list of schema versions, and stamps out a static variable with an array containing those versions, plus an rstest_reuse template with multiple attributes, one for each schema version we want to test against.

I was thinking that we could modify this macro to add another set of case attributes on the template test, one for each schema version except the last one, that runs up scripts until version i+1, then runs the corresponding down script to revert to version i. We'd need to write a new function to pass to the case attribute, change the macro arguments to something like `($( $i:literal ),* , $i_last:literal)`, and change the attribute repetitions like so.

```rust
$(#[case(ephemeral_datastore_max_schema_version($i))])*
#[case(ephemeral_datastore_max_schema_version($i_last))]
$(#[case(ephemeral_datastore_max_schema_version_with_revert($i))])*
```

(we may want to throw in a static assert that `std::cmp::max(&SUPPORTED_SCHEMA_VERSIONS) == SUPPORTED_SCHEMA_VERSIONS[SUPPORTED_SCHEMA_VERSIONS.length - 1]` for good measure)

We can't meaningfully run unit tests in a way that exercises `.down.sql` scripts while the code only supports one schema version (presumably the latest), so I think it'll still be important to have a separate smoke test that applies every up migration, then reverts using every down migration, to confirm they at least work on an empty database.

**Comment by inahga** [comment]
That sounds... really solid! I should be able to cook up a POC today.

One concern: let me try to think about the situation where we have multiple supported schema versions. Suppose we have versions `x` and `y` where `x < y`.
- I write functionality that requires schema `y`.
- I write a unit test that tests functions that use schema `y`.
- My unit test of course fails when tested against schema `x`.

This suggests then any functions that require schema `y` must always check that the database is running on schema `y` before proceeding, and the unit tests must check that the function is checking the schema version.

This sounds perfectly fine to me, but is this consistent with how we plan on achieving compatibility with multiple schemas, i.e. all functions depending on a newer schema must change their behavior based on schema version?

**Comment by divergentdave** [comment]
So far we've treated support for different schema versions as all-or-nothing, both through running unit tests against all supported versions, and by checking the database against `SUPPORTED_SCHEMA_VERSIONS` in `Datastore::new()`. I think the plan is that we'd have to shrink the list of supported schema versions before writing queries that impose hard requirements on, say, new columns, rather than having individual methods check the version. Supporting multiple schema versions in one release of the code just gives us flexibility to roll out creation of those new columns, before we start depending on them in a future release.

**Comment by inahga** [comment]
> > that down scripts return the schema to the state it was in before running the corresponding up script (perhaps by dumping the schema before and after the up/down migration pair and diffing?)
> 
> As for schema validation, there's a few options:
> 
> 1. Dump tables/indices/types/etc. by hand and compare.
> 2. Find a dedicated tool for diff'ing pg schemas.
> 3. It's probably worth trying to see if diff <(pgdump db_normal) <(pgdump db_downgraded) would work?
> 
> Option 2 sounds like the better idea--but I'm still researching what the best options are. Open to input if anyone knows a good tool off-hand.

I was able to test the [`pg_dump` approach](https://github.com/divviup/janus/commit/08453d6d5a4373fd8ac15cb7121b93f61efccf57). It *almost* works, surprisingly the schema dumps are deterministic. However, the table column order is presented in creation order. For instance, if we drop a column in an up migration, and re-add the column in the down migration, the result is that the column will be the last one in the pg_dump, and `diff` will see this as a difference. This is seen in version 2
```diff
332d331
<     out_share bytea,
334c333,334
<     last_prep_step bytea
---
>     last_prep_step bytea,
>     out_share bytea
```
I think absent some brutal hacks, that means `pg_dump` is out and we probably have to explore other tools.

Of the third party tools I've found, https://github.com/djrobstep/migra is the most likely candidate. It has the most coverage of postgresql features. Drawbacks are it requires a Python environment to run, or use of its docker image. It also doesn't appear to a ton of industry usage and maintenance appears to be slow. The other candidates I've found are abandoned or incomplete. I would approach this tool with the mindset that we might have to support/patch it ourselves if we want to keep using it.

I'm curious though if we feel we would get enough value out of the newly added tests (syntax checking on down migrations, janus tests against down migrations) that it's not worth adding a new dependency?

For clarity: the question this proposed test is answering is "does the down migration revert the schema to its previous version _exactly_"?

**Comment by branlwyd** [comment]
I'm OK not maintaining column order, only the set of columns.

This is open for discussion, but:
* I'd prefer not to introduce another tool (e.g. migra) to perform schema-equivalence checking if we can find another solution.
* If we can't find a solution, I think it's acceptable to lean only on our functional tests, which _should_ exercise that the schema is in working order. (This wouldn't catch e.g. a vestigial nullable column being left around after a downgrade, but I think this doesn't matter much. Maybe I'm overlooking some important schema diff that wouldn't be caught by functional tests?)

That said, before we do the operational work to introduce a new tool, WDYT about using queries against [Postgres' system catalogs](https://www.postgresql.org/docs/current/catalogs.html) to retrieve a programmatic list of various schema entities, which we can then compare in an order-agnostic way. I think this is likely to be more maintainable than learning & maintaining a tool, since we already have tooling to easily spin up a database at a given migration version (which, IIUC, we'd need to do to dump schema anyway).

BTW, I really like Migra, specifically its ability to introspect two complete schema definitions and produce a diff -- I wish our migration story had been able to do something like this, as determining the "final" schema from a sequence of small migrations can be challenging.

**Comment by inahga** [comment]
Let me take some time to try doctoring up some comparisons of the system catalog +/- `information_schema`. My kneejerk aversion to it is achieving completeness of verification could take a considerable amount of code, pushing it out of scope for this issue. I desire completeness because we don't want the test to give us a false sense of security (i.e. missing potentially crucial changes). This is without having actually tried it first, so let me see if I'm overstating it.

> (This wouldn't catch e.g. a vestigial nullable column being left around after a downgrade, but I think this doesn't matter much. Maybe I'm overlooking some important schema diff that wouldn't be caught by functional tests?)

My main thought was something like a new a new (broken) constraint not being rolled back on the down migration. I think something like that would be caught by code review + the new tests (i.e. likely failing to remove a constraint is a syntax/tx runtime error).

---

I do wish pgdump worked, since it's the most complete serialization of the db schema you can get. I was vaguely thinking of some logic where the developer accepts diffs in the schema. If a diff is detected, the test fails and the developer is alerted. If they accept the difference, i.e. because it's benign, they can add to a list of accepted diffs, perhaps by shasum. Exchanges some additional work for completeness of schema analysis. Feels quite atypical/hacky though. Can prototype that too to see how it feels.

**Comment by divergentdave** [comment]
I think comparing the schema of two live database catalogs against each other will share a lot of challenges with the question of declarative database schemas, or database migration tools with tighter integrations than simply running SQL scripts, as they all require maintaining a separate copy of the Postgres object model. (As an example, I recall reading about some tool that depended on reading the catalog but produced the wrong answer when dealing with a UNIQUE constraint that covered more than one column, because it used a SQL query that didn't join the various catalog tables correctly. I'm afraid I can't find a link to it again.) Comparing two catalogs could be easier than the other cases, because we could get some value out of even a rudimentary implementation that only handles a few classes of objects (i.e. tables, indices, and columns) and ignores most of their attributes. For starters, note that each object is identified by an auto-incrementing `oid`, where increments are shared across the whole database, so we'll have to resolve all foreign key oid columns as references to other objects, and not compare any oid values directly.

I agree that the combination of functional tests on rolled-back databases and the smoke test in which we revert back to version 0 should get us most of what we want. That'll flush out practically all syntax/data problems between queries and the schema.

**Comment by inahga** [comment]
Here's the prototype for querying the system tables: https://github.com/divviup/janus/commit/dab22aeba82569115b0811a23631ea4d524f2d5c. For now, it just queries the `pg_class` table which includes tables, views, indices, other things. Pardon the horrible iterator chains, this was experimentation in how to make the data more presentable.

It's definitely fragile to do it this way--I had to spend quite a bit of time discovering that the `relnatts` is asynchronously updated when columns are dropped (by vacuum? not at all? it's not actually documented...),. I'm thinking to expand it out to cover the other entities like indices, columns, constraints, etc. the effort will exceed the value it brings.

Another experiment was vaguely how we can get pg_dump comparisons to work https://github.com/divviup/janus/compare/main...inahga/test-db-down-migrations-by-schema, i.e. keep a static collection of acceptable schema differences. This is less fragile, but hacky and requires manual intervention to work around false positives. I do like the spirit of this technique--would it be helpful to have a test/utility that easily dumps the schema differences to aid in writing downgrade migrations?

**Comment by branlwyd** [comment]
Thinking about this some more, I think we should rely on our functional tests, i.e. do not inspect the database schema. We should file an issue to improve on this later if we want to.

The SQL query looks fairly straightforward, but owning all of Postgres' behavior around their system catalogs (e.g. the `relnatts` issue) is unpalatable. OTOH, third-party tools will require some operational maintenance, and it is hard for us to guarantee that their output is correct (without learning enough about them that we may as well own the SQL queries ourselves), especially during DB version upgrades.

Given that we get 95% of the way there with our functional tests, I think that's a good stopping point for now. (I did realize our functional tests are not likely to notice a missing index -- this would lead to a performance hit, which is not tested and would be unlikely to be noticed by a human given the scale of our datastore tests.)
